### ISSUE TITLE
self-tests #94 && #95 failing

### ISSUE THREAD

**Issue by pabeni** [multipath-tcp/mptcp_net-next#265]
running in a loop:

while ./mptcp_join.sh 94 95; do : ; done

I see sporadic failures - VM with 3 vCPUs, debug build.

```
Created /tmp/tmp.GhtUQmr445 (size 1024 KB) containing data sent by client
Created /tmp/tmp.zHXDJjzfeb (size 1024 KB) containing data sent by server
copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 4)
copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 4)
 client exit code 2, server 0

netns ns1-KDjXGP socket stat for 10094:
Failed to find cgroup2 mount
TcpPassiveOpens                 2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      95                 0.0
TcpOutRsts                      6                  0.0
TcpExtTCPPureAcks               21                 0.0
TcpExtTCPLossProbes             1                  0.0
TcpExtTCPOrigDataSent           81                 0.0
TcpExtTCPDelivered              80                 0.0
MPTcpExtMPCapableSYNRX          1                  0.0
MPTcpExtMPCapableACKRX          1                  0.0
MPTcpExtMPJoinSynRx             1                  0.0
MPTcpExtMPJoinAckRx             1                  0.0
MPTcpExtDataCsumErr             2                  0.0
MPTcpExtMPFailTx                2                  0.0
MPTcpExtMPFastcloseTx           2                  0.0
MPTcpExtMPRstTx                 2                  0.0

netns ns2-KDjXGP socket stat for 10094:
Failed to find cgroup2 mount
TcpActiveOpens                  2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      55                 0.0
TcpExtTCPPureAcks               6                  0.0
TcpExtTCPOrigDataSent           31                 0.0
TcpExtTCPDelivered              18                 0.0
MPTcpExtMPCapableSYNTX          1                  0.0
MPTcpExtMPCapableSYNACKRX       1                  0.0
MPTcpExtMPTCPRetrans            1                  0.0
MPTcpExtMPJoinSynAckRx          1                  0.0
MPTcpExtOFOQueueTail            5                  0.0
MPTcpExtOFOQueue                6                  0.0
MPTcpExtOFOMerge                5                  0.0
MPTcpExtMPFailRx                2                  0.0
MPTcpExtMPRstRx                 2                  0.0
./mptcp_join.sh: line 1210: [: [{"total acts":1},{"actions":[{"order":0 pedit ,"control_action":{"type":"pipe"}keys 1
 	 index 1 ref 1 bind 1,"installed":3062,"last_used":3016
	 key #0  at 148: val ff000000 mask ffffffff
1: integer expression expected
095 MP_FAIL MP_RST                       syn[ ok ] - synack[ ok ] - ack[ ok ]
                                         sum[ ok ] - csum  [ ok ]
                                         ftx[fail] got 2 MP_FAIL[s] TX expected 1
 - failrx[fail] got 2 MP_FAIL[s] RX expected 1
Server ns stats
TcpPassiveOpens                 2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      95                 0.0
TcpOutRsts                      6                  0.0
TcpExtTCPPureAcks               21                 0.0
TcpExtTCPLossProbes             1                  0.0
TcpExtTCPOrigDataSent           81                 0.0
TcpExtTCPDelivered              80                 0.0
MPTcpExtMPCapableSYNRX          1                  0.0
MPTcpExtMPCapableACKRX          1                  0.0
MPTcpExtMPJoinSynRx             1                  0.0
MPTcpExtMPJoinAckRx             1                  0.0
MPTcpExtDataCsumErr             2                  0.0
MPTcpExtMPFailTx                2                  0.0
MPTcpExtMPFastcloseTx           2                  0.0
MPTcpExtMPRstTx                 2                  0.0
Client ns stats
TcpActiveOpens                  2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      55                 0.0
TcpExtTCPPureAcks               6                  0.0
TcpExtTCPOrigDataSent           31                 0.0
TcpExtTCPDelivered              18                 0.0
MPTcpExtMPCapableSYNTX          1                  0.0
MPTcpExtMPCapableSYNACKRX       1                  0.0
MPTcpExtMPTCPRetrans            1                  0.0
MPTcpExtMPJoinSynAckRx          1                  0.0
MPTcpExtOFOQueueTail            5                  0.0
MPTcpExtOFOQueue                6                  0.0
MPTcpExtOFOMerge                5                  0.0
MPTcpExtMPFailRx                2                  0.0
MPTcpExtMPRstRx                 2                  0.0
                                         rtx[fail] got 2 MP_RST[s] TX expected 1
 - rstrx [fail] got 2 MP_RST[s] RX expected 1
Server ns stats
TcpPassiveOpens                 2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      95                 0.0
TcpOutRsts                      6                  0.0
TcpExtTCPPureAcks               21                 0.0
TcpExtTCPLossProbes             1                  0.0
TcpExtTCPOrigDataSent           81                 0.0
TcpExtTCPDelivered              80                 0.0
MPTcpExtMPCapableSYNRX          1                  0.0
MPTcpExtMPCapableACKRX          1                  0.0
MPTcpExtMPJoinSynRx             1                  0.0
MPTcpExtMPJoinAckRx             1                  0.0
MPTcpExtDataCsumErr             2                  0.0
MPTcpExtMPFailTx                2                  0.0
MPTcpExtMPFastcloseTx           2                  0.0
MPTcpExtMPRstTx                 2                  0.0
Client ns stats
TcpActiveOpens                  2                  0.0
TcpEstabResets                  2                  0.0
TcpInSegs                       34                 0.0
TcpOutSegs                      55                 0.0
TcpExtTCPPureAcks               6                  0.0
TcpExtTCPOrigDataSent           31                 0.0
TcpExtTCPDelivered              18                 0.0
MPTcpExtMPCapableSYNTX          1                  0.0
MPTcpExtMPCapableSYNACKRX       1                  0.0
MPTcpExtMPTCPRetrans            1                  0.0
MPTcpExtMPJoinSynAckRx          1                  0.0
MPTcpExtOFOQueueTail            5                  0.0
MPTcpExtOFOQueue                6                  0.0
MPTcpExtOFOMerge                5                  0.0
MPTcpExtMPFailRx                2                  0.0
MPTcpExtMPRstRx                 2                  0.0

                                         itx[ ok ] - infirx[ ok ]

1 failure(s) has(ve) been detected:
	- 95: MP_FAIL MP_RST
```
Note: the bad outout:

```
./mptcp_join.sh: line 1210: [: [{"total acts":1},{"actions":[{"order":0 pedit ,"control_action":{"type":"pipe"}keys 1
 	 index 1 ref 1 bind 1,"installed":3062,"last_used":3016
	 key #0  at 148: val ff000000 mask ffffffff
1: integer expression expected
```
is due to old/buggy tc version, does not affect the tests themself - e.g. I see the failure even with a recent tc version/without the above bash splat

**Comment by geliangtang** [comment]
A kernel log is attached.

[repeat-log-32245-204.dmesg.txt](https://github.com/multipath-tcp/mptcp_net-next/files/8309261/repeat-log-32245-204.dmesg.txt)

**Comment by matttbe** [comment]
From the last meeting:

      - It looks like we got 2 checksum failures on 2 different failures:
          - it triggers 2 checksum failures → 2 MP Fail
          - if it is the case, could we make sure we only "break" one subflow?
          - but we should because the 'iptables' rule should mark only one packet (can be one packet before TSO)
          - so only one packet should be corrupted on one single sublow

**Comment by matttbe** [comment]
@pabeni recently sent a new patch:

New patches for t/upstream:
- b841c3f765af: net/sched: act_pedit: really ensure the skb is writable
- Results: 2767792d035c..8ae619c5e009 (export)

New patches for t/upstream-net:
- b841c3f765af: net/sched: act_pedit: really ensure the skb is writable
- Results: eefc441cb5ab..8b01b4ca3343 (export-net)

I quote:

> This almost solves issues/265 here. I'm still getting some rare failure with MPTcpExtMPFailTx==0: sometimes the transfer completes before we are able to use the 2nd/failing link. The relevant fix is a purely seft-test one

I guess we leave this ticket opens to have a fix for that, right?
Should we have more data to transfer? Wait for an event? Reduce BW? Introduce delays/losses on the first link to force using the 2nd subflow?

**Comment by geliangtang** [comment]
The old tc version bad outout issue is fixed by my commit 2770b28b85fb ("selftests: mptcp: fix a mp_fail test warning").

**Comment by pabeni** [comment]
> @pabeni recently sent a new patch:
> 
> New patches for t/upstream:
> 
>     * [b841c3f](https://github.com/multipath-tcp/mptcp_net-next/commit/b841c3f765af1e7af669ce7c1b406a60668d4b4e): net/sched: act_pedit: really ensure the skb is writable
> 
>     * Results: [2767792](https://github.com/multipath-tcp/mptcp_net-next/commit/2767792d035c53b26cbdf47176beba4703acc406)..[8ae619c](https://github.com/multipath-tcp/mptcp_net-next/commit/8ae619c5e009f5a1e0e03be0e9085edc7dfe801e) (export)
> 
> 
> New patches for t/upstream-net:
> 
>     * [b841c3f](https://github.com/multipath-tcp/mptcp_net-next/commit/b841c3f765af1e7af669ce7c1b406a60668d4b4e): net/sched: act_pedit: really ensure the skb is writable
> 
>     * Results: [eefc441](https://github.com/multipath-tcp/mptcp_net-next/commit/eefc441cb5abe84dc544a1859994a25c8bb443f8)..[8b01b4c](https://github.com/multipath-tcp/mptcp_net-next/commit/8b01b4ca3343862e8dd3167fc0233e7bf905e2b6) (export-net)
> 
> 
> I quote:
> 
> > This almost solves issues/265 here. I'm still getting some rare failure with MPTcpExtMPFailTx==0: sometimes the transfer completes before we are able to use the 2nd/failing link. The relevant fix is a purely seft-test one
> 
> I guess we leave this ticket opens to have a fix for that, right? Should we have more data to transfer? Wait for an event? Reduce BW? Introduce delays/losses on the first link to force using the 2nd subflow?

IMHO we are good, all the relevant fixes are in the export branch and I do not see failure anymore
